Add timeout reset on progress notifications

Ozamatash opened the pull request:
# Add timeout reset on progress  notifications

Closes #87 

## Motivation and Context
This change addresses the need for better handling of long-running operations in the MCP protocol. Previously, requests would time out according to a fixed timeout, even if the server was actively sending progress notifications. This could lead to premature timeouts for valid long-running operations.

The implementation allows servers to indicate that an operation is still in flight by sending progress notifications, which will reset the timeout. To prevent indefinite operations, a configurable maximum total timeout is also provided.

## How Has This Been Tested?
- Basic timeout reset on progress notification
- Maximum total timeout enforcement
- Multiple sequential progress notifications
- Default timeout behavior without progress
- Edge cases around timeout boundaries

## Breaking Changes
No breaking changes. The implementation is fully backwards compatible:
- New options (`resetTimeoutOnProgress` and `maxTotalTimeout`) are optional
- Default timeout behavior remains unchanged
- Existing progress notification handling is preserved

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context

jspahrsummers commented on src/shared/protocol.test.ts:
```diff
@@ -62,6 +62,182 @@ describe("protocol tests", () => {
   });
+
+  test("should reset timeout when progress notification is received", async () => {
+    jest.useFakeTimers();
```
Can you please put this into a `beforeEach` and `useRealTimers` into an `afterEach`? (Consider using a `describe` to group with the below tests too.)

Right now, errors in the test could cause timers not to be restored.

jspahrsummers commented on src/shared/protocol.test.ts:
```diff
@@ -62,6 +62,182 @@ describe("protocol tests", () => {
+      progress: 50,
+      total: 100,
+    });
+
```
Presumably we need to advance the timers again here? Otherwise this doesn't seem to really be testing that the timeout got extended.

jspahrsummers commented on src/shared/protocol.test.ts:
```diff
@@ -62,6 +62,182 @@ describe("protocol tests", () => {
+      onprogress: onProgressMock,
+    });
+
+    // Advance time beyond maxTotalTimeout
+    jest.advanceTimersByTime(150);
+
+    // Send progress notification after maxTotalTimeout
+    if (transport.onmessage) {
+      transport.onmessage({
+        jsonrpc: "2.0",
+        method: "notifications/progress",
+        params: {
+          progressToken: 0,
+          progress: 50,
+          total: 100,
+        },
+      });
+    }
```
We should test that sending a progress notification doesn't _reset_ the max total timeout, which AFAICT this doesn't exercise—because the timeout already happened by the time the progress notification is sent.

jspahrsummers commented on src/shared/protocol.ts:
```diff
@@ -64,6 +64,20 @@ export type RequestOptions = {
+
+  /**
+   * Maximum total time (in milliseconds) to wait for a response, even if progress notifications are received.
+   * Only used when resetTimeoutOnProgress is true.
```
It might be a bit confusing that this property only has an effect when another is set. Any reason not to always apply it?

jspahrsummers commented on src/shared/protocol.ts:
```diff
@@ -149,6 +175,49 @@ export abstract class Protocol<
+    const totalElapsed = Date.now() - info.startTime;
+    if (info.maxTotalTimeout && totalElapsed >= info.maxTotalTimeout) {
+      this._timeoutInfo.delete(messageId);
+      cancel(new McpError(
+        ErrorCode.RequestTimeout,
+        "Maximum total timeout exceeded",
+        { maxTotalTimeout: info.maxTotalTimeout, totalElapsed }
+      ));
```
TIOLI: I think the flow here is a bit confusing, and IMO it'd be cleaner to throw this as an exception here, rather than chain callbacks.

jspahrsummers reviewed (changes requested):
Thanks for working on this! ✨

Ozamatash commented:
1. Test improvements:
   - Moved timer setup/cleanup into beforeEach/afterEach hooks.
   - Added timer advancement after progress notification to properly test timeout extension
   - Fixed maxTotalTimeout test to verify progress notifications don't reset after timeout

2. Code changes:
   - Made maxTotalTimeout always apply, independent of resetTimeoutOnProgress
   - Changed error handling to throw exception directly instead of callbacks.
   - Added try-catch in _onprogress to properly handle timeout errors

jspahrsummers commented on src/shared/protocol.ts:
```diff
@@ -360,7 +359,12 @@ export abstract class Protocol<
     if (this._timeoutInfo.has(messageId) && responseHandler) {
-      if (!this._resetTimeout(messageId, (reason) => responseHandler(reason as Error))) {
+      try {
+        if (!this._resetTimeout(messageId)) {
+          return;
+        }
```
IIUC this means that _not_ having a timeout handler will skip calling the progress handler. Is that intentional? (Sorry for missing it in the first review!)

jspahrsummers reviewed (changes requested):
Thanks, looking good! Just one remaining question—sorry for overlooking it the first time.

jspahrsummers reviewed (approved):
Great, thanks!

GLips commented:
This looks interesting for [an issue I'm working on](https://github.com/GLips/Figma-Context-MCP/issues/3#issuecomment-2679907339). Are there updated docs for this feature? I can't tell directly from the PR how I might use it—I don't think I have direct access to the `protocol.request` function in my implementation.

Ozamatash commented:
On the server side you need to be sending notification messages and on the client side this feature needs to be enabled in the RequestOptions. This commit wasn't included in the latest release so it's not yet in the official npm package if I understand correctly. Basically mcp clients like Cursor still need to enable this.

GLips commented:
Appreciate the response @Ozamatash—I'm not clear on how to send notification messages from looking at the typescript SDK documentation or the tests in your PR, but sounds like if I hold tight support (and docs) for this will percolate through the ecosystem. Thanks for the good work!
